### ISSUE TITLE
[Feature] Stabilize DeepSeek V4 device-side lm_head and decouple TP from EP

### ISSUE THREAD

**Issue by sjduan** [hw-native-sys/pypto-lib#696]
## Summary

Support a stable **device-side DeepSeek V4 `lm_head` path** for serving, with **TP=4** and with `lm_head` TP decoupled from the decode/prefill EP world size.

The original goal was to validate `lm_head.py --tp 4` as a standalone kernel. That is still necessary, but it is no longer sufficient. In serving-style runs, the important path is:

```text
decode_fwd -> post-final-norm hidden_out -> device-side lm_head -> logits / token selection
```

We have seen cases where standalone `lm_head` can pass, but the composed `decode_fwd + lm_head` device-side path may stall or hang at runtime. This issue should track both the correctness/performance bring-up and the distributed synchronization / rank-mapping work needed to make this path reliable in serving.

## Problem Statement

Current DeepSeek V4 forward drivers primarily expose a hidden-state contract: `decode_fwd.py` / `prefill_fwd.py` return post-final-norm hidden states, and serving can run `lm_head` on the host side.

For production serving we also need a device-side path. Two blockers remain:

1. **Runtime stability of composed `decode_fwd + lm_head`**

   A standalone TP-sharded `lm_head` kernel is not enough. When `lm_head` is composed after `decode_fwd`, the distributed publish/gather path can sometimes stall in service-like runs. The fix should make the device-side `lm_head` communication protocol robust when it is launched after the full decode graph, not just as an isolated unit test.

2. **TP / EP decoupling**

   The current experimental composition can assume `tp == ep` / `TP_SIZE == N_RANKS`. Serving should be able to run the decode forward with one EP configuration and run the final vocabulary projection with a different TP group, for example:

   ```text
   decode_fwd EP=8 -> hidden_out owner ranks
   lm_head TP=4 -> 4 vocabulary shards
   ```

   The implementation should not rely on a one-to-one mapping between EP ranks and LM-head TP ranks.

## Motivation

DeepSeek V4 serving eventually needs a device-side final projection path so that normalized hidden states can be projected to vocabulary logits without falling back to CPU/host `lm_head`.

This matters for:

- reducing host/device transfer in serving;
- enabling a clean device-side token selection path;
- supporting TP-sharded vocabulary projection for the large DeepSeek V4 vocab;
- making the final projection path consistent with the rest of distributed decode.

## Current State

Relevant current behavior:

- `decode_fwd.py` / `prefill_fwd.py` emit post-final-norm hidden states in the default path.
- `models/deepseek/v4/lm_head.py` already has a TP-sharded device kernel and supports `--tp 4` at the interface level.
- Standalone TP=4 validation is useful, but the serving-critical path is composition with `decode_fwd`.
- Earlier experimental composition used `tp == ep` as a simplifying assumption.
- In composed runs, the logits routing / gather synchronization path can be more fragile than standalone `lm_head`, so the issue should explicitly require runtime stability under `decode_fwd + lm_head`.

### Phase 1: Harden standalone device-side `lm_head` TP=4

Validate `models/deepseek/v4/lm_head.py --tp 4` as an independent device kernel:

- compile PASS;
- runtime PASS;
- golden compare PASS;
- full logits gather correctness across 4 TP shards;
- coverage of the TP=4 vocab-tail path.

### Phase 2: Compose `decode_fwd + lm_head` on device

Add an optional device-side composition path:

```text
decode_fwd(...) -> hidden_out
lm_head_tp(hidden_out, lm_head_weight) -> logits / token output
```

Requirements:

- The default hidden-output ABI may remain available for host-side fallback.
- The composed path must not hang or stall under normal serving-style runtime.
- Distributed communication ordering must be explicit enough that `finish_logits` cannot observe incomplete remote shard writes.
- The PR should document the required window buffers, output shape, and whether it emits full logits, last-token logits, or token ids.

### Phase 3: Decouple LM-head TP from decode EP

Generalize the rank mapping so that LM-head TP is not assumed to equal decode EP.

The implementation should define:

- EP rank group: ranks that run decode attention/MoE and own `hidden_out` rows;
- TP rank group: ranks that own `lm_head_weight` vocab shards and produce logits shards;
- how hidden rows are published from EP owners to TP ranks;
- how logits or token-selection output is gathered back;
- what configurations are supported initially.

Suggested first target configurations:

- `ep=4, tp=4` as the simplest composed baseline;
- `ep=8, tp=4` as the first true TP/EP decoupling case.

### Phase 4: Serving-oriented output mode

After full-logits composition is stable, evaluate whether serving should output:

- full `[T, vocab]` logits;
- last-token logits only;
- top-k / argmax token ids only.

This should be decided based on memory, window size, and serving requirements.

## Acceptance Criteria

- `lm_head.py --tp 4` passes compile/runtime/golden compare on NPU.
- `decode_fwd + device-side lm_head` passes runtime without stalls/hangs.
- The composed path is tested repeatedly enough to rule out intermittent synchronization failure.
- TP=4 vocab-tail path is covered.
- At least one `tp == ep` configuration passes, preferably `ep=4,tp=4`.
- At least one `tp != ep` configuration passes, preferably `ep=8,tp=4`.
- The implementation documents the rank mapping and window-buffer ownership for EP ranks and TP ranks.
- The default host-side `lm_head` / hidden-output path remains available unless intentionally replaced.

## Non-goals For The First Milestone

- Do not require TP=8/TP=16 bring-up.
- Do not require sampling/top-p/top-k implementation; logits or greedy token output is enough.
- Do not require real-weight end-to-end text generation as the first acceptance gate.
- Do not require removing the host-side fallback path immediately.

## Related Context

This request follows the current hidden-state forward contract where DeepSeek V4 L3 forward kernels return post-final-norm hidden states for serving-side `lm_head`. The new requirement is to restore and stabilize a **device-side** `lm_head` option, including the realistic composed serving path and TP/EP decoupling.

**Comment by zhangqi-chen** [comment]
Closing as done — verified against main `e058d747`.

Both blockers this issue tracked are resolved in `models/deepseek_v4_flash_mtp/`:

1. **Composed `decode_fwd` + device-side `lm_head`** — `decode_fwd.py:731-738` calls `lm_head_with_sampling` inside the forward, driven by the `lm_head_hidden_window` / `lm_head_logits_window` publish-gather pair with explicit `*_done` flags and a comm epoch, so the projection runs on device as part of the decode graph rather than on the host. Landed via #711, #835 and #117.

2. **TP / EP decoupling** — `decode_fwd.py:1861-1862` exposes `--ep` (EP world size) and `--tp` (LM-head TP world size) as independent arguments, and the rank mapping is `my_rank // LM_HEAD_TP_SIZE * LM_HEAD_TP_SIZE` + `my_rank % LM_HEAD_TP_SIZE`, i.e. TP groups are derived from the EP rank space rather than assuming `tp == ep`.

Remaining `lm_head` work is performance, not bring-up, and belongs in its own issue.
